cmake/Subdirectories.cmake

```diff
@@ -153,24 +160,112 @@ endfunction()
 function(query_gem_paths_from_external_subdirs output_gem_dirs gem_names registered_external_subdirs)
     if (gem_names)
-        foreach(gem_name IN LISTS gem_names)
+        foreach(gem_name_with_version_specifier IN LISTS gem_names)
             unset(gem_path)
+
+            # Remove the version specifier from the gem name before fetching properties
+            o3de_get_name_and_version_specifier(${gem_name_with_version_specifier} gem_name spec_op spec_version)
+
             get_property(gem_optional GLOBAL PROPERTY ${gem_name}_OPTIONAL)
-            o3de_find_gem_with_registered_external_subdirs(${gem_name} gem_path "${registered_external_subdirs}")
+            get_property(gem_path GLOBAL PROPERTY "@GEMROOT:${gem_name}@")
+            if(NOT gem_path)
+                if(NOT O3DE_DISABLE_GEM_DEPENDENCY_RESOLUTION)
+                    message(WARNING "Failed to find resolved gem path for gem '${gem_name_with_version_specifier}'")
+                    # Make a best effort attempt to find the most compatible gem
+                    o3de_find_most_compatible_gem_with_registered_external_subdirs(${gem_name_with_version_specifier} gem_path "${registered_external_subdirs}")
+                else()
+                    o3de_find_gem_with_registered_external_subdirs(${gem_name_with_version_specifier} gem_path "${registered_external_subdirs}")
+                endif()
+
+                if(gem_path)
+                    set_property(GLOBAL PROPERTY "@GEMROOT:${gem_name}@" "${gem_path}")
+                endif()
+            endif()
+
             if (gem_path)
                 list(APPEND gem_dirs ${gem_path})
             elseif(NOT gem_optional)
-                list(JOIN registered_external_subdirs "\n" external_subdirs_formatted)
+                # Sort the list so it is easier to search visually
+                list(SORT registered_external_subdirs COMPARE NATURAL CASE INSENSITIVE ORDER ASCENDING)
+                # Indent the text to be easier to read. If the indent is removed CMake will add an
+                # additional newline automatically because "non-indented text is formatted in 
+                # line-wrapped paragraphs delimited by newlines"
+                list(JOIN registered_external_subdirs "\n  " external_subdirs_formatted)
                 message(SEND_ERROR "The gem \"${gem_name}\""
-                " could not be found in any gem.json from the following list of registered external subdirectories:\n"
-                "${external_subdirs_formatted}")
+                " could not be found in any gem.json from the following list of registered external subdirectories:"
+                "\n  ${external_subdirs_formatted}")
                 break()
             endif()
         endforeach()
     endif()
     set(${output_gem_dirs} ${gem_dirs} PARENT_SCOPE)
 endfunction()
 
+#! Use cmake.py to get a resolved list of gem names and paths for this object (engine or project)
```

Comment:
so how is the cmake system used in conjunction with cmake.py?
Does cmake.py run cmake commands while injecting necessary values into it?

Reply:
1. `Subdirectores.cmake` runs `cmake.py`, giving it the path to the engine or project.  `cmake.py` does not run any cmake commands, only Python
2. `cmake.py` determines all the gems an engine or project uses and runs dependency resolution on them
3. `cmake.py`  writes the output as a CMake list to a file 
4. `Subdirectories.cmake` loads the file and writes the resolved gem name to gem path results into CMake global property variables, so that when any logic wants to know the path to some gem, it get's the resolved path to the gem from that global property.

This way, if you have say 100 versions of PopcornFX gems on your computer, CMake can determine what single gem path should be used for PopcornFX in your project.